src/coord/src/coord.rs

```diff
@@ -3027,55 +3162,35 @@ impl<S: Append + 'static> Coordinator<S> {
+        let mut write_rx = None;
+        let mut internal_command_tx = None;
 
         if let EndTransactionAction::Commit = action {
             if let Some(ops) = txn.into_ops() {
                 match ops {
-                    TransactionOps::Writes(inserts) => {
-                        // Although the transaction has a wall_time in its pcx, we use a new
-                        // coordinator timestamp here to provide linearizability. The wall_time does
-                        // not have to relate to the write time.
-                        let (timestamp, advance_to) = self.get_and_step_local_write_ts();
-                        let mut appends: HashMap<GlobalId, Vec<Update<Timestamp>>> = HashMap::new();
-
-                        for WriteOp { id, rows } in inserts {
+                    TransactionOps::Writes(mut writes) => {
+                        for WriteOp { id, .. } in &writes {
                             // Re-verify this id exists.
                             let _ = self.catalog.try_get_entry(&id).ok_or_else(|| {
                                 CoordError::SqlCatalog(CatalogError::UnknownItem(id.to_string()))
                             })?;
-                            // This can be empty if, say, a DELETE's WHERE clause had 0 results.
-                            if rows.is_empty() {
-                                continue;
-                            }
-                            let updates = rows
-                                .into_iter()
-                                .map(|(row, diff)| Update {
-                                    row,
-                                    diff,
-                                    timestamp,
-                                })
-                                .collect::<Vec<_>>();
-                            appends.entry(id).or_default().extend(updates);
                         }
 
-                        match appends.len() {
-                            1 => {
-                                let (id, updates) = appends.into_element();
-                                self.dataflow_client
-                                    .storage_mut()
-                                    .append(vec![(id, updates, advance_to)])
-                                    .await
-                                    .unwrap();
-                            },
-                            0 => {},
-                            _ => unreachable!("multi-table write transaction should fail immediately when the second table is added to the transaction"),
+                        // `rows` can be empty if, say, a DELETE's WHERE clause had 0 results.
+                        writes.retain(|WriteOp { rows, .. }| !rows.is_empty());
+                        let (tx, rx) = oneshot::channel();
+                        write_rx = Some(rx);
+                        if self.pending_writes.is_empty() {
+                            internal_command_tx = Some(self.internal_cmd_tx.clone());
```

Comment:
The message send can happen right here, doesn't need to be below.

Reply:
Yes, good point. I moved it here.